Material utility operations namespace

ezrayst opened the pull request:
Response to #587. Computations and tests for:
- [x] mean p 
- [x] deviatoric stress
- [x] deviatoric q
- [x] J2
- [x] J3 
- [x] Lode angle theta
- [x] dp_dsigma
- [x] dq_dsigma
- [x] dj2_dsigma
- [x] dj3_dsigma
- [x] dtheta_dsigma

Upcoming work for next PR:
- [ ] Refactor NorSand
- [ ] Refactor Modified Cam Clay
- [ ] Refactor Mohr Coulomb

kks32 commented on include/materials/tensor.h:
```diff
@@ -0,0 +1,66 @@
+namespace tensor {
+
+//! Compute mean stress p
+//! \param[in] stress Stress in Voigt notation where positive is tension
```
Why is tension positive? In geotech applications most of our stresses are compressive

ezrayst replied:
I am open to discussion. My thought is to keep this tension positive for the purpose of other functions like computing deviatoric stress. In most geotech application, then users can multiply by -1 separately in material functions?

kks32 replied:
The MPM code is mainly for Geotech applications, also I don't think we need to explicity define positive / negative

ezrayst replied:
OK let's get only `mean_p` positive, but the rest to be the same definition

ezrayst replied:
Done

kks32 commented on include/materials/tensor.h:
```diff
@@ -0,0 +1,66 @@
+#include "Eigen/Dense"
+
+namespace mpm {
+namespace tensor {
```
We may have to reconsider the namespace naming, none of the operations are tensorial operations. Maybe just `material` or `invariants`

ezrayst replied:
I am fine with `material`, `invariants` might seem a little too specific because this may grow bigger than just computing the `invariants`. Maybe others can comment what they think too.

ezrayst replied:
But then the name of the files would be `material` which would be the same as the current class `material`

bodhinandach replied:
I vote for `material_utility`, basically, it can also contain other than invariants if future constitutive laws need it.

ezrayst replied:
Done :) unless others have better names

kks32 commented on include/materials/tensor.tcc:
```diff
@@ -0,0 +1,226 @@
+//! Compute mean stress p
+inline double mpm::tensor::compute_mean_p(
```
This should be const everything

kks32 replied:
Almost all functions should be `const` everything

ezrayst replied:
Alright, I will `const` everything.

kks32 commented on include/materials/tensor.tcc:
```diff
@@ -0,0 +1,226 @@
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute mean p in tension positive
+  double mean_p = 1. / 3. * (stress(0) + stress(1) + stress(2));
```
return this instead of creating a variable.

ezrayst replied:
Noted

kks32 commented on include/materials/tensor.tcc:
```diff
@@ -0,0 +1,226 @@
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute mean p in tension positive
+  double mean_p = compute_mean_p(stress);
```
const

ezrayst replied:
Done

kks32 commented on include/materials/tensor.tcc:
```diff
@@ -0,0 +1,226 @@
+  double dtheta_dR = -1.0 / 3.0;
+
+  // Update when J2 is non zero
+  if (abs(j2) > 1.0E-6) {
```
Why 1.0E-6?

ezrayst replied:
We can change them of course, but I just chose this as a safe-guard. Note that since we are talking about Pa as unit for stress, it is rarely that we have stress so small that it would trigger this if condition. It usually only happens when we have zero stresses all throughout.

Should we change them to 1.E-16 instead?

tianchiTJ commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  double lode_angle = (1. / 3.) * acos(lode_angle_val);
+  if (lode_angle > M_PI / 3.) lode_angle = M_PI / 3.;
+  if (lode_angle < -0.) lode_angle = 0.;
+
```
I think no need to check it again if you have limited the lode_angle_val in [-1,1]

ezrayst replied:
I see, but it's not harmful to have right?

bodhinandach replied:
If conds are expensive, I second @tianchiTJ. It's always a good practice to minimize if :)

kks32 replied:
If we have checked and restricted the values to be between -1 and 1, why should we check again?

ezrayst replied:
Ok, I have removed it.

tianchiTJ commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+
+  // Compute dq / dsigma
+  Eigen::Matrix<double, 6, 1> dq_dsigma = Eigen::Matrix<double, 6, 1>::Zero();
+  if (abs(deviatoric_q) > 1.E-6) {
```
Do you think this tolerance is too large? Do we need to define a same tolerance for all denominator?

ezrayst replied:
I don't mind using 1.E-15 actually, but it will blow up more right?

bodhinandach replied:
maybe -8 or -9 to be in the middle?

kks32 replied:
We should pass a default tolerance value

ezrayst replied:
OK

bodhinandach replied:
@kks32 isn't better to create `mpm::tolerance()` so that by default all tolerance is the same by default? Or we can utilize `std::numeric_limits<double>::epsilon()`

ezrayst replied:
I have used the latter actually `std::numeric_limits<double>::epsilon()`

bodhinandach commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+        const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute deviatoric stress
+  Eigen::Matrix<double, 6, 1> dev_stress = compute_deviatoric_stress(stress);
```
const?

kks32 replied:
Why do we write to `deviatoric_stress` and then copy it to dj2/dsigma, why can't we assign directly?

ezrayst replied:
Will do so

bodhinandach commented on include/materials/material_utility.h:
```diff
@@ -0,0 +1,68 @@
+
+//! Compute mean stress p
+//! \param[in] stress Stress in Voigt notation where positive is tension
+inline const double compute_mean_p(const Eigen::Matrix<double, 6, 1>& stress);
```
Can we include the details that p is assumed to be positive in compression here too?

ezrayst replied:
I see, ok

kks32 commented on include/materials/material_utility.h:
```diff
@@ -0,0 +1,68 @@
+inline const double compute_mean_p(const Eigen::Matrix<double, 6, 1>& stress);
+
+//! Compute deviatoric stress
+//! \param[in] stress Stress in Voigt notation where positive is tension
```
Please change this to positive is compression

ezrayst replied:
Wait, the input stress is always positive tension

kks32 replied:
I'm really sorry, I got this mixed up. You are right, stresses we get are positive in tension. Let's be consistent and have mean stress be also positive in tension. Sorry about that!! I was incorrectly looking at the magnitude not a component.

kks32 commented on include/materials/material_utility.h:
```diff
@@ -0,0 +1,68 @@
+
+//! Compute mean stress p
+//! \param[in] stress Stress in Voigt notation where positive is tension
+inline const double compute_mean_p(const Eigen::Matrix<double, 6, 1>& stress);
```
Shall we rename the function as `p` and drop the `compute_` part from all the function names?

kks32 replied:
I'm in two minds about the namespace `material_utility` instead of just `materials` because every time we use it, we have to do `mpm::material_utility::compute_` instead of `mpm::material::mean_p`, which would be more succinct to read.

ezrayst replied:
I understand. Should the filename be changed too? I will keep the filename as is for now

bodhinandach replied:
material is a bit off I guess. I am still for material utility. Btw, @kks32 for the namespace are using CamelCase or snake_case?

kks32 commented on include/materials/material_utility.h:
```diff
@@ -0,0 +1,68 @@
+
+#include <cmath>
+
+#include "Eigen/Dense"
```
```suggestion
#include "data_types.h"
```
We can use that as the default include

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+//! Compute mean stress p
+inline const double mpm::material_utility::compute_mean_p(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+  // Compute mean p in compresion positive
```
Could we move this comment to the top, as this isn't adding any additional value here.

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+    mpm::material_utility::compute_deviatoric_stress(
+        const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute mean p in tension positive
```
Could we please have compression positive?

ezrayst replied:
OK, it has been changed

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  const double mean_p = -1.0 * compute_mean_p(stress);
+
+  // Compute deviatoric by subtracting volumetric part
+  Eigen::Matrix<double, 6, 1> dev_stress = stress;
```
Could you please rename to deviatoric stress

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  Eigen::Matrix<double, 6, 1> dev_stress = compute_deviatoric_stress(stress);
+
+  // Compute J3
+  double j3 = (dev_stress(0) * dev_stress(1) * dev_stress(2)) -
```
const double

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  // Compute J2 from
+  const double j2 = compute_j2(stress);
+
+  const double deviatoric_q = std::sqrt(3 * j2);
```
No need to define a new variable, return `std::sqrt(3*j2)

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+
+//! Compute Lode angle
+inline const double mpm::material_utility::compute_lode_angle(
+    const Eigen::Matrix<double, 6, 1>& stress) {
```
I'd suggest passing a default tolerance value

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+
+  // Compute Lode angle value
+  double lode_angle_val = 0.0;
+  if (abs(j2) > 1.0E-6) {
```
```suggestion
  if (std::abs(j2) > tolerance) {
```

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+    mpm::material_utility::compute_dp_dsigma(
+        const Eigen::Matrix<double, 6, 1>& stress) {
+
+  Eigen::Matrix<double, 6, 1> dp_dsigma = Eigen::Matrix<double, 6, 1>::Zero();
```
Please use the data type function `zero()`

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+//! Compute derivative of q in terms of stress sigma
+inline const Eigen::Matrix<double, 6, 1>
+    mpm::material_utility::compute_dq_dsigma(
+        const Eigen::Matrix<double, 6, 1>& stress) {
```
I recommend passing a default argument value for tolerance

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  Eigen::Matrix<double, 6, 1> dev_stress = compute_deviatoric_stress(stress);
+
+  // Compute dq / dsigma
+  Eigen::Matrix<double, 6, 1> dq_dsigma = Eigen::Matrix<double, 6, 1>::Zero();
```
Please use data_type function zero

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  const double deviatoric_q = compute_deviatoric_q(stress);
+
+  // Compute deviatoric stress
+  Eigen::Matrix<double, 6, 1> dev_stress = compute_deviatoric_stress(stress);
```
do we need to specify type here? const auto?

ezrayst replied:
I can use const auto. I think it's better

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  // Compute dq / dsigma
+  Eigen::Matrix<double, 6, 1> dq_dsigma = Eigen::Matrix<double, 6, 1>::Zero();
+  if (abs(deviatoric_q) > 1.E-6) {
+    dq_dsigma(0) = 3. / 2. / deviatoric_q * dev_stress(0);
```
Could you please use parenthesis continuous division is confusing to read.

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  dev3(1) = dev_stress(4);
+  dev3(2) = dev_stress(2);
+
+  Eigen::Matrix<double, 6, 1> dj3_dsigma = Eigen::Matrix<double, 6, 1>::Zero();
```
use data_type function zero

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  const double j2 = compute_j2(stress);
+
+  // Compute deviatoric stress
+  const Eigen::Matrix<double, 6, 1> dev_stress =
```
const auto

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  const double j3 = compute_j3(stress);
+
+  // Compute dj2_dsigma
+  const Eigen::Matrix<double, 6, 1> dj2_dsigma = compute_dj2_dsigma(stress);
```
const auto

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  const Eigen::Matrix<double, 6, 1> dj2_dsigma = compute_dj2_dsigma(stress);
+
+  // Compute dj3_dsigma
+  const Eigen::Matrix<double, 6, 1> dj3_dsigma = compute_dj3_dsigma(stress);
```
const auto

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  double dtheta_dR = -1.0 / 3.0;
+
+  // Update when J2 is non zero
+  if (abs(j2) > 1.0E-6) {
```
std::abs(j2) > tolerance

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,230 @@
+  // Update when J2 is non zero
+  if (abs(j2) > 1.0E-6) {
+    // Update R
+    R = j3 / 2.0 * std::pow(j2 / 3.0, -1.5);
```
please only use lower case letters for variables

ezrayst replied:
OK

kks32 commented on tests/materials/material_utility_test.cc:
```diff
@@ -0,0 +1,213 @@
+#include <cmath>
```
You don't need to include all of these again if you are including material_utility.h

ezrayst replied:
OK

kks32 commented on tests/materials/material_utility_test.cc:
```diff
@@ -0,0 +1,213 @@
+#include "material_utility.h"
+
+//! \brief Check material_utility class for 3D case
+TEST_CASE("Material utility is checked for 3D case", "[material_utility][3D]") {
```
Why specifically is it a 3D case?

ezrayst replied:
I guess there is no 3D or 2D in this case, everything is generalized as 3D. I will delete them.

ezrayst commented:
I think I have addressed all the comments. Now, the namespace is called `material` but the file is still called `material_utility`. 

Throwing some ideas here: `material_functions`, `material_operations`, `materials` (with s) if you want something short.

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,221 @@
+  // Compute J2 from
+  const double j2 = mpm::material::j2(stress);
+
+  return (std::sqrt(3 * j2));
```
Could we just write out a formula instead of a function call?

ezrayst replied:
That defeats the purpose of having this PR right, where we want to reduce repetition of codes, and easier to maintain.

kks32 replied:
function calls are expensive, and this is a single line function for J2, if this gets called a lot, it's a lot more expensive!

ezrayst replied:
Another way is to clump the different functions together, so all the other functions that need j2 would be in the same call. I feel if we repeat the single line function, it would be hard to maintain.

ezrayst replied:
I am still on the fence about this. Does it mean that we have to copy the function (one line) for both j2 and j3?

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,221 @@
+inline const Eigen::Matrix<double, 6, 1> mpm::material::dp_dsigma(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  Eigen::Matrix<double, 6, 1> dp_dsigma;
```
Using set zero will be 2 steps every time, the way you had it was how we want it, but instead use the `zero<6>()` to initialize them to zeros.

kks32 replied:
This function is available in `data_types.h`, you may want to specialize for `6`

ezrayst replied:
I see! Give me a moment

ezrayst replied:
I updated the `data_types.h` but is this how I use it?
`Eigen::Matrix<double, 6, 1> dp_dsigma = mpm::zero<6>();`? (It does not work)

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,221 @@
+
+  // Declare R as zero to avoid division by zero J2
+  // R is defined as R = cos(3 theta)
+  double r = 0.0;
```
Why are we declaring r here? Couldn't we declare it line 204?

ezrayst replied:
We can't because r is zero when j2 is zero (division by zero) so the same as the next three variables, we need to declare here and then update

kks32 replied:
But you are not using `r` anywhere else outside the loop starting in line 202. Or am I missing something?

ezrayst replied:
Sorry, we can!

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,221 @@
+    dr_dj2 *= std::pow(j2, -2.5);
+    dr_dj3 *= std::pow(j2, -1.5);
+    // Update derivative of theta in terms of R, check for sqrt of zero
+    if (std::abs(1 - r * r) < std::numeric_limits<double>::epsilon()) {
```
const double factor = (std::abs(1 - r * r) < std::numeric_limits<double>::epsilon()) ?  std::numeric_limits<double>::epsilon() : (1 - r*r);
dtheta_dr = -1.0/(3*sqrt(factor));

ezrayst replied:
Understood

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,221 @@
+
+  // Compute Lode angle value
+  double lode_angle_val = 0.0;
+  if (std::abs(j2) > std::numeric_limits<double>::epsilon()) {
```
Isn't this such a small value? Why can't we set a default tolerance, which constitutive modelers can modify?

ezrayst replied:
Yes, it is such a small value and not likely to happen except almost zero stress (j2 is a squared term). But not all constitutive models have tolerance.

kks32 replied:
If we have a tolerance value as a default argument, then modelers needing a tolerance can use it, others can ignore. Any reason, why we shouldn't do this?

ezrayst replied:
I don't mind actually, I will add that.

bodhinandach replied:
I second @kks32

ezrayst replied:
I have overloaded the functions

kks32 replied:
Set a default argument, please don't overload the functions

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,221 @@
+  const Eigen::Matrix<double, 6, 1> dtheta_dsigma =
+      dtheta_dr * ((dr_dj2 * dj2_dsigma) + (dr_dj3 * dj3_dsigma));
+
+  return dtheta_dsigma;
```
There is no need for `dtheta_dsigma`, return directly

ezrayst commented:
@kks32, how about the namespace name? should we keep them as `material` and the filenames as `material_utility`?

kks32 commented:
I'd go with `materials` for namespace and keep the filename as `material_utilities.h`?

ezrayst commented:
Understood.

tianchiTJ commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,271 @@
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute mean p
+  const double p = mpm::materials::p(stress);
```
I found some paper use p=0.5*(sigma1+sigma3) and q=0.5*(sigma1-sigma3) for 2D. I think q is ok, and the current equation can represent the same result, but the p may be different. Sorry I'm not sure about this part, do you know which one is better?

kks32 replied:
Hi @tianchiTJ those are MIT definition of mean and deviatoric stress. I suggest we stick to octohedral stress definition in 3D.

tianchiTJ replied:
OK, got it

kks32 commented on include/data_types.h:
```diff
@@ -16,19 +16,25 @@ using Index = unsigned long long;
 template <typename Ttype>
```
If you are not using this file, there is no need to change this.

kks32 commented on include/materials/material_utility.h:
```diff
@@ -29,15 +29,12 @@ inline const double j3(const Eigen::Matrix<double, 6, 1>& stress);
-                               const double tolerance);
+inline const double lode_angle(
+    const Eigen::Matrix<double, 6, 1>& stress,
+    const double tolerance = std::numeric_limits<double>::epsilon());
```
Please don't const double in input argument or anything of size double or lower: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-in

kks32 commented on include/materials/material_utility.h:
```diff
@@ -59,16 +56,12 @@ inline const Eigen::Matrix<double, 6, 1> dj2_dsigma(
 //! \param[in] tolerance Default tolerance value specified by user
 inline const Eigen::Matrix<double, 6, 1> dtheta_dsigma(
-    const Eigen::Matrix<double, 6, 1>& stress, const double tolerance);
+    const Eigen::Matrix<double, 6, 1>& stress,
```
Please avoid const here in tolerance

kks32 commented on include/materials/material_utility.h:
```diff
@@ -0,0 +1,71 @@
+
+//! Compute mean stress p (tension positive)
+//! \param[in] stress Stress in Voigt notation where positive is tension
+inline const double p(const Eigen::Matrix<double, 6, 1>& stress);
```
Please avoid returning by const value for built-in data types. 
```you are promising that the original variable (inside the function frame) will not be modified. And again, because you’re returning it by value, it’s copied so the original value could never be modified via the return value. For built-in types, it doesn’t matter whether you return by value as a const, so you should avoid confusing the client programmer and leave off the const when returning a built-in type by value.```

For more info: https://www.linuxtopia.org/online_books/programming_books/thinking_in_c++/Chapter08_014.html

kks32 replied:
Keep const for `Eigen` data types

ezrayst replied:
I see. Understood.

kks32 commented on include/materials/material_utility.tcc:
```diff
@@ -0,0 +1,209 @@
+//! Compute mean stress p (compression positive)
```
```suggestion
//! Compute mean stress p (tension is positive)
```
